qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/FeatureUpgradeIT.java

```diff
@@ -0,0 +1,101 @@
+
+            Request bulk = new Request("POST", "/_bulk");
+            bulk.addParameter("refresh", "true");
+            bulk.setJsonEntity("{\"index\": {\"_index\": \"test_index_old\"}}\n" +
+                "{\"f1\": \"v1\", \"f2\": \"v2\"}\n");
+            client().performRequest(bulk);
+
+            // start a async reindex job
+            Request reindex = new Request("POST", "/_reindex");
+            reindex.setJsonEntity(
+                "{\n" +
+                    "  \"source\":{\n" +
+                    "    \"index\":\"test_index_old\"\n" +
+                    "  },\n" +
+                    "  \"dest\":{\n" +
+                    "    \"index\":\"test_index_reindex\"\n" +
+                    "  }\n" +
+                    "}");
+            reindex.addParameter("wait_for_completion", "false");
+            Map<String, Object> response = entityAsMap(client().performRequest(reindex));
+            String taskId = (String) response.get("task");
+
+            // wait for task
+            Request getTask = new Request("GET", "/_tasks/" + taskId);
+            getTask.addParameter("wait_for_completion", "true");
+            client().performRequest(getTask);
+
+            // make sure .tasks index exists
+            Request getTasksIndex = new Request("GET", "/.tasks");
+            getTasksIndex.setOptions(expectVersionSpecificWarnings(v -> {
+                v.current(systemIndexWarning);
+                v.compatible(systemIndexWarning);
+            }));
+            getTasksIndex.addParameter("allow_no_indices", "false");
+
+            getTasksIndex.setOptions(expectVersionSpecificWarnings(v -> {
+                v.current(systemIndexWarning);
+                v.compatible(systemIndexWarning);
+            }));
+            assertBusy(() -> {
+                try {
+                    assertThat(client().performRequest(getTasksIndex).getStatusLine().getStatusCode(), is(200));
+                } catch (ResponseException e) {
+                    throw new AssertionError(".tasks index does not exist yet");
+                }
+            });
+
+        } else if (CLUSTER_TYPE == ClusterType.UPGRADED) {
+            // check results
+            assertBusy(() -> {
+                Request clusterStateRequest = new Request("GET", "/_migration/system_features");
+                XContentTestUtils.JsonMapView view = new XContentTestUtils.JsonMapView(
+                    entityAsMap(client().performRequest(clusterStateRequest)));
+
+                List<Map<String, Object>> features = view.get("features");
+                Map<String, Object> feature = features.stream()
+                    .filter(e -> "tasks".equals(e.get("feature_name")))
+                    .findFirst()
+                    .orElse(Collections.emptyMap());
+
+                assertThat(feature.size(), equalTo(4));
+                assertThat(feature.get("minimum_index_version"), equalTo(UPGRADE_FROM_VERSION.toString()));
```

Comment:
I suspect we could do some logic on the old and new versions and check the `(NO_)UPGRADE_NEEDED` bit as well, if the old version is a different major than the new version.